### ISSUE TITLE
`multiSchema` support for MySQL

### ISSUE THREAD

**Issue by Jolg42** [prisma/prisma#16943]
`multiSchema` is done for PostgreSQL, CockroachDB and SQLServer but we miss support for MySQL.

**Comment by floelhoeffel** [comment]
I'll convert this to the main issue tracking support for MySQL for multiSchema

**Comment by ludufre** [comment]
Version 5 was released and this feature here is still open. Please don't forget this, Prisma is very good and I would really like to implement it in our main project.

**Comment by bentleysayer23** [comment]
Any updates on this? We are wanting to switch asap but have been blocked by this for a long time

**Comment by William-McGonagle** [comment]
@Jolg42 

Any update on this issue? I know it's still open but we have a MySQL database and we're looking to functionally shard it soon so it can better scale with us.

**Comment by Ustice** [comment]
I'm looking forward to seeing what can be done in this space given the restructuring of how Prisma is handling preview features, given that MySQL is listed as a first-class database in terms of support.

**Comment by thelinuxlich** [comment]
Hoping to see this come to life in 2025 :)

**Comment by scavlev** [comment]
While undocumented - it used to work before version 5.15.0 even with mysql. We are still using 5.14.0 version with multiSchema + mysql and are pretty much stuck with it as it looks like this commit https://github.com/prisma/prisma-engines/commit/ae0473d812da9ce31f8c62740c937afb9623f2dc has broken it.

**Comment by thelinuxlich** [comment]
Is it supported now that multi schemas are out of preview?

**Comment by NanNanCHN** [comment]
> While undocumented - it used to work before version 5.15.0 even with mysql. We are still using 5.14.0 version with multiSchema + mysql and are pretty much stuck with it as it looks like this commit [prisma/prisma-engines@ae0473d](https://github.com/prisma/prisma-engines/commit/ae0473d812da9ce31f8c62740c937afb9623f2dc) has broken it.

what about this https://www.prisma.io/blog/organize-your-prisma-schema-with-multi-file-support, did you try it?

**Comment by thelinuxlich** [comment]
Not the same thing

**Comment by FGoessler** [comment]
### Important Update On This Feature Request

As we are working towards bringing the `multiSchema` feature from preview to general availability, we also had another deeper look into supporting multiple schemas for MySQL.

What we want to achieve with this feature is that users can spread out their models across schemas in their database via a simple `@@schema("<schema-name>")` attribute on the model in the Prisma schema file and query models across schemas through a single `PrismaClient` instance ([docu](https://www.prisma.io/docs/orm/prisma-schema/data-model/multi-schema)).

> [!NOTE]
> Please do not confuse this with having multiple Prisma schema files!
> You can read more about using multiple Prisma schema files [here](https://www.prisma.io/docs/orm/prisma-schema/overview/location#multi-file-prisma-schema).

Unfortunately, MySQL does not natively support this layer of abstraction as it treats databases and schemas as synonyms - contrary to the SQL standard and e.g. Postgres and SQL Server.


> [!IMPORTANT]
> **Hence we have concluded that it is not technically feasible with justifiable efforts and we will not pursue this feature for MySQL for the foreseeable future.**

We understand this may be disappointing to some users. However, demand for this feature has been comparatively low. If you have strong demand for this feature please open a new issue describing your specific use-case and motivation.

### Alternatives To Use

If you have the need to connect to multiple databases on your MySQL server you can create separate `PrismaClient` specifications with separate schema files and output directories. These clients can be then used alongside each other to query the databases independently. 

---

### Why This Is Not Technically Feasible For Us:

#### Context:

- The SQL standard specifies a “three tiered hierarchy” of: `database` → `schema` → `table`
- MySQL, unlike PostgreSQL, does **not** follow this hierarchy. In MySQL `DATABASE` and `SCHEMA` are effectively synonyms. In fact `CREATE DATABASE foo` and `CREATE SCHEMA foo` result in the same operation.
- Prisma’s migration engine internally uses a so called “shadow database” to perform introspections and schema diffing ([read more about it here](https://www.prisma.io/docs/orm/prisma-migrate/understanding-prisma-migrate/shadow-database)). The shadow database needs an isolated namespace to create ephemeral tables that can be created and destroyed adhoc by commands such as  `prisma migrate dev`.
- Currently that level of isolation of the shadow database is ensured by creating a fresh database via a `CREATE DATABASE <generated-tmp-shaddow-db-name>`  statement.

#### Problem:

- Because `CREATE SCHEMA` and `CREATE DATABASE` are equivalent in MySQL, attempting to isolate schemas within a shadow database does not work.
- Any `CREATE SCHEMA` call during a shadow DB operation escapes the intended isolation boundary and creates a new top-level database on the server.
- This can interfere with the user’s actual databases and potentially cause destructive side effects.
- ⇒ In short, **we cannot guarantee the isolation required for safe shadow database operations when using multiple schemas in MySQL**.

### Why We Can’t Work Around It

While theoretical workarounds exist, they all come with **serious limitations**:

1. Dynamic SQL rewriting
    
    Attempt to detect and rewrite schema-scoped SQL to use globally unique schema names during shadow database operations.
    
    - Highly fragile: migrations may contain arbitrary SQL that’s hard to rewrite safely.
    - Increases the risk of subtle or destructive errors.
    - Weakens isolation guarantees.
2. Require a dedicated MySQL instance
    
    Ask users to provide a fully isolated MySQL server exclusively for the shadow database.
    
    - Poor user experience.
    - High potential for misconfiguration and footguns (e.g., using the same instance unintentionally).

This limitation also impacts our internal testing setup. Without schema-level isolation, we cannot safely run parallel tests on shared MySQL instances, making test setup more complex and brittle.

### Additional Considerations:

We also encountered issues when testing with Vitess and PlanetScale-based MySQL setups. These systems use schemas internally for sharding and branching, and may rename schemas in metadata tables needed for introspection. This adds another layer of fragility and complexity, making robust support even more challenging.
